server/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java

```diff
@@ -100,6 +100,10 @@ private SearchHits buildSearchHits(SearchContext context, Profiler profiler) {
         FetchContext fetchContext = new FetchContext(context);
         SourceLoader sourceLoader = context.newSourceLoader();
 
+        PreloadedSourceProvider sourceProvider = new PreloadedSourceProvider();
+        PreloadedFieldLookupProvider fieldLookupProvider = new PreloadedFieldLookupProvider();
+        context.getSearchExecutionContext().setLookupProviders(sourceProvider, ctx -> fieldLookupProvider);
```

Comment:
Would it be possible to add a unit test as well? I guess it really has to do with how top hits interacts with the fetch phase?

Reply:
Yes, unit tests for this will be tricky.  We don't really have unit tests for any of the FetchPhase at the moment as the context structures that it uses are so heavy.